### ISSUE TITLE
Ubernetes e2e flakes

### ISSUE THREAD

**Issue by madhusudancs** [kubernetes/kubernetes#26762]
I am opening this master/umbrella issue to track Ubernetes e2e test flakes. I think this issue is sufficient right now. But as we make progress we might have to fork this into sub-issues.

This issue is not just about the e2e tests themselves but also includes infrastructure related flakes such as bringing up the clusters, bringing up the federation control plane, etc.

Here is one flake that I am seeing from yesterday:

```shell
$ FEDERATION=true E2E_ZONES="us-central1-f" KUBE_GCE_ZONE="us-central1-f" go run hack/e2e.go -v --up
...
...
...
NAME                                     LOCATION       SCOPE  BASE_INSTANCE_NAME                       SIZE  TARGET_SIZE  INSTANCE_TEMPLATE                           AUTOSCALED
madhusudancs-us-central1-f-minion-group  us-central1-f  zone   madhusudancs-us-central1-f-minion-group        3            madhusudancs-us-central1-f-minion-template
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Waiting for group to become stable, current operations: creating: 3
Group is stable
INSTANCE_GROUPS=madhusudancs-us-central1-f-minion-group
NODE_NAMES=madhusudancs-us-central1-f-minion-group-5wk9 madhusudancs-us-central1-f-minion-group-6ys7 madhusudancs-us-central1-f-minion-group-l1vs
Using master: madhusudancs-us-central1-f-master (external IP: <REDACTED_IP>)
Waiting up to 300 seconds for cluster initialization.

  This will continually check to see if the API for kubernetes is reachable.
  This may time out if there was some uncaught error during start up.

...................................................................................................................................................Cluster failed to initialize within 300 seconds.
2016/06/02 14:18:06 e2e.go:218: Error running up: exit status 1
2016/06/02 14:18:06 e2e.go:214: Step 'up' finished in 8m0.261564635s
2016/06/02 14:18:06 e2e.go:114: Error starting e2e cluster. Aborting.
exit status 1
```

cc @colhom @nikhiljindal @colhom @kubernetes/sig-cluster-federation @jianhuiz @mfanjie

**Comment by quinton-hoole** [comment]
@madhusudancs I saw similar symptoms today, but it turns out the I'd reached the quota on my project.  Do you have sufficient quota (specifically I had the default quota of 24 cores, and the federation requires more than that).  I trivially increased my quota with an increase request, which got automatically approved within a few seconds.

**Comment by madhusudancs** [comment]
@quinton-hoole I have a higher quota limit in the region where I am spinning up a federation. Btw, this is a single cluster federation.

**Comment by quinton-hoole** [comment]
For the record, my three clusters came up fine, but then I got the following.  Will debug some more in the morning:

```
$ FEDERATION=true E2E_ZONES="us-central1-a us-central1-b us-central1-f" KUBE_GCE_ZONE="us-central1-f" FEDERATION_PUSH_REPO_BASE="gcr.io/quinton-etcd-testing" KUBERNETES_PROVIDER="gce" federation/cluster/federated-up.sh 
namespace "federation-e2e" created
No resources found
service "federation-apiserver" created
attempting to get federation-apiserver loadbalancer hostname (1 / 30)
attempting to get federation-apiserver loadbalancer hostname (2 / 30)
attempting to get federation-apiserver loadbalancer hostname (3 / 30)
attempting to get federation-apiserver loadbalancer hostname (4 / 30)
attempting to get federation-apiserver loadbalancer hostname (5 / 30)
attempting to get federation-apiserver loadbalancer hostname (6 / 30)
Found federation-apiserver host at 104.197.87.156
deployment "federation-apiserver" created
secret "federation-apiserver-secrets" created
deployment "federation-controller-manager" created
Waiting for federation-apiserver to be running...(phase= Pending)
...
Waiting for federation-apiserver to be running...(phase= Pending)
federation-apiserver pod is not running! giving up.
```

**Comment by quinton-hoole** [comment]
PS:  It looks like the images are not where they should be:

```
kubectl --server=https://146.148.37.247 get pods --namespace federation-e2e
NAME                                             READY     STATUS             RESTARTS   AGE
federation-apiserver-3663518660-nwge3            1/2       ImagePullBackOff   0          5m
federation-controller-manager-2288429398-98142   0/1       ImagePullBackOff   0          5m
```

**Comment by quinton-hoole** [comment]
I'm pretty sure that this is operator error on my part.

I ran this:

```
FEDERATION=true E2E_ZONES="us-central1-a us-central1-b us-central1-f" KUBE_GCE_ZONE="us-central1-f" FEDERATION_PUSH_REPO_BASE="gcr.io/quinton-etcd-testing" go run hack/e2e.go -v --up
```

and after all three clusters came up cleanly, I got this...

``` 
k8s.io/kubernetes/federation/cluster/common.sh: line 75: KUBERNETES_PROVIDER: unbound variable
2016/06/03 00:09:08 e2e.go:218: Error running up: exit status 1
2016/06/03 00:09:08 e2e.go:214: Step 'up' finished in 23m42.294835356s
2016/06/03 00:09:08 e2e.go:114: Error starting e2e cluster. Aborting.
exit status 1
```
So (stupidly), rather than tearing the whole lot down and starting over, I ran this:

```
$ FEDERATION=true E2E_ZONES="us-central1-a us-central1-b us-central1-f" KUBE_GCE_ZONE="us-central1-f" FEDERATION_PUSH_REPO_BASE="gcr.io/quinton-etcd-testing" KUBERNETES_PROVIDER="gce" federation/cluster/federated-up.sh -v
```
and got this:

```
namespace "federation-e2e" created
No resources found
service "federation-apiserver" created
attempting to get federation-apiserver loadbalancer hostname (1 / 30)
attempting to get federation-apiserver loadbalancer hostname (2 / 30)
attempting to get federation-apiserver loadbalancer hostname (3 / 30)
attempting to get federation-apiserver loadbalancer hostname (4 / 30)
attempting to get federation-apiserver loadbalancer hostname (5 / 30)
attempting to get federation-apiserver loadbalancer hostname (6 / 30)
Found federation-apiserver host at 104.197.87.156
deployment "federation-apiserver" created
secret "federation-apiserver-secrets" created
deployment "federation-controller-manager" created
Waiting for federation-apiserver to be running...(phase= Pending)
...
Waiting for federation-apiserver to be running...(phase= Pending)
federation-apiserver pod is not running! giving up.
```

I'm busy tearing it all down and starting from scratch, with the correct env vars.

**Comment by madhusudancs** [comment]
Please don't close this issue until we are reasonably confident that this isn't a flake. If it works once that doesn't mean it is not a flake. Also, this is an umbrella issue.

**Comment by madhusudancs** [comment]
@quinton-hoole Did you also run `federation/cluster/federation-push.sh` before running `federation/cluster/federation-up.sh` or `hack/e2e.go --up`? Also, please see the discussion here - https://github.com/kubernetes/kubernetes/issues/26656

**Comment by quinton-hoole** [comment]
My apologies.  I had no intention of closing this issue.  I must have pressed the wrong button :-(

**Comment by quinton-hoole** [comment]
And thanks - my federation is now up and healthy:

Waiting for federation-apiserver to be running...(phase= Pending)
Waiting for federation-apiserver to be running...(phase= Running)
federation-apiserver pod is running!
Waiting for federation-controller-manager to be running...(phase= Running)
federation-controller-manager pod is running!
cluster "federated-cluster" set.
user "federated-cluster" set.
context "federated-cluster" set.
Wrote config for federated-cluster to /Users/quinton/.kube/config

kubectl get pods --namespace federation-e2e
NAME                                             READY     STATUS    RESTARTS   AGE
federation-apiserver-608821220-rsid0             2/2       Running   0          2m
federation-controller-manager-3887574383-h40yw   1/1       Running   0          2m

**Comment by quinton-hoole** [comment]
And the e2e test for cluster registration passed!

[It] should allow creation of cluster api objects
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-apiserver.go:63
Jun  3 09:02:42.316: INFO: Creating cluster object: federation-e2e-gce-us-central1-a (https://104.197.87.156)
Jun  3 09:02:42.746: INFO: Creating cluster object: federation-e2e-gce-us-central1-b (https://104.197.70.145)
Jun  3 09:02:42.812: INFO: Creating cluster object: federation-e2e-gce-us-central1-f (https://23.236.59.255)

...

Ran 1 of 312 Specs in 98.316 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 311 Skipped PASS

Ginkgo ran 1 suite in 1m38.613311042s
Test Suite Passed

**Comment by quinton-hoole** [comment]
But my cluster controller is spewing errors, unable to list clusters.  @nikhiljindal @mfanjie @madhusudancs @jianhuiz 

E0603 16:11:08.140689       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/cluster/clustercontroller.go:111: Failed to list *federation.Cluster: Get federation-apiserver:443?resourceVersion=0: unsupported protocol scheme "federation-apiserver"

**Comment by nikhiljindal** [comment]
Looks like we need to replace `--master={{.FEDERATION_APISERVER_DEPLOYMENT_NAME}}:443` by `--master=https://{{.FEDERATION_APISERVER_DEPLOYMENT_NAME}}:443` here: https://github.com/kubernetes/kubernetes/blob/d2ea7a29ab710f10a0f12a7c2d06cea662b20bf5/federation/manifests/federation-controller-manager-deployment.yaml#L21.

I havent tried it. @quinton-hoole can you try and see if that fixes the problem?

**Comment by bprashanth** [comment]
Random sidenote given that this is a e2e umbrella. A while back I noticed ubernetes lite e2es weren't collecting logs from all nodes: https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-ubernetes-lite/4407/artifacts/

**Comment by quinton-hoole** [comment]
Thanks @bprashanth.  Good catch.  That should be pretty easy to fix.  I
imagine that the e2e log collector is getting confused about the multi-zone
cluster.

**Comment by quinton-hoole** [comment]
Thanks @nikhiljindal, that improved things, but cluster controller is still unable to connect to API server (see below).  I'm going to head in to the office now, will continue debugging shortly:

```
kubectl log federation-controller-manager-3548753978-1gv2b --namespace federation-e2e
...
E0603 18:01:44.649876       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/service/servicecontroller.go:232: Failed to list *federation.Cluster: Get https://federation-apiserver:443/apis/federation/v1alpha1/clusters?resourceVersion=0: dial tcp 10.0.240.175:443: getsockopt: no route to host
E0603 18:02:02.661837       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/cluster/clustercontroller.go:111: Failed to list *federation.Cluster: Get https://federation-apiserver:443/apis/federation/v1alpha1/clusters?resourceVersion=0: dial tcp: i/o timeout
E0603 18:02:08.667227       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/service/servicecontroller.go:231: Failed to list *api.Service: Get https://federation-apiserver:443/api/v1/services?resourceVersion=0: dial tcp 10.0.240.175:443: i/o timeout
```

The service exists, although those ports look weird:

```
$ kubectl get services --namespace federation-e2e
NAME                   CLUSTER-IP     EXTERNAL-IP      PORT(S)   AGE
federation-apiserver   10.0.240.175   146.148.79.202   443/TCP   7m
$ kubectl describe services --namespace federation-e2e
Name:			federation-apiserver
Namespace:		federation-e2e
Labels:			app=federated-cluster
Selector:		app=federated-cluster,module=federation-apiserver
Type:			LoadBalancer
IP:			10.0.240.175
LoadBalancer Ingress:	146.148.79.202
Port:			https	443/TCP
NodePort:		https	32152/TCP
Endpoints:		10.180.1.2:443
Session Affinity:	None
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  8m		8m		1	{service-controller }			Normal		CreatingLoadBalancer	Creating load balancer
  7m		7m		1	{service-controller }			Normal		CreatedLoadBalancer	Created load balancer

```

**Comment by quinton-hoole** [comment]
Oh never mind, the ports are fine.  It was just my kubectl output that got mushed in my terminal window.

**Comment by quinton-hoole** [comment]
And occasionally, this error:

```
E0603 19:09:26.005217       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/service/servicecontroller.go:232: Failed to list *federation.Cluster: Get https://federation-apiserver:443/apis/federation/v1alpha1/clusters?resourceVersion=0: x509: failed to load system roots and no roots provided
```

**Comment by quinton-hoole** [comment]
OK, the problem seems to be as simple as this (from API server logs):

I0603 19:10:34.136146       1 logs.go:41] http: TLS handshake error from 10.240.0.13:61288: tls: client offered an unsupported, maximum protocol version of 301
I0603 19:12:53.040150       1 logs.go:41] http: TLS handshake error from 10.240.0.11:50186: remote error: bad certificate

**Comment by quinton-hoole** [comment]
It seems that we need the correct certificates installed on the controller manager.

In the mean time I tried setting up unsecured, non-https access to API server, but that has other problems.  Apparently the generic apiserver only listens on localhost for unsecured access:

I0603 20:35:02.441500       1 genericapiserver.go:690] Serving securely on 0.0.0.0:443
I0603 20:35:03.222121       1 genericapiserver.go:703] Using self-signed cert (/var/run/kubernetes/apiserver.crt, /var/run/kubernetes/apiserver.key)
I0603 20:35:03.222214       1 genericapiserver.go:734] Serving insecurely on 127.0.0.1:8080

So, not surprisingly, controller manager can't connect:

E0603 20:43:05.903765       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/service/servicecontroller.go:231: Failed to list *api.Service: Get http://federation-apiserver:8080/api/v1/services?resourceVersion=0: dial tcp 10.0.114.72:8080: getsockopt: connection refused
E0603 20:43:05.903783       1 reflector.go:216] k8s.io/kubernetes/federation/pkg/federation-controller/service/servicecontroller.go:232: Failed to list *federation.Cluster: Get http://federation-apiserver:8080/apis/federation/v1alpha1/clusters?resourceVersion=0: dial tcp 10.0.114.72:8080: getsockopt: connection refused

**Comment by mfanjie** [comment]
#26694 should fix the unsecured connection issue

**Comment by nikhiljindal** [comment]
Thanks @mfanjie!

I am fixing the secured connection in https://github.com/kubernetes/kubernetes/pull/26819. I need to update the federated-up.sh script after that to auto create the secret.

We should have e2e tests for both secured and unsecured connections.

**Comment by quinton-hoole** [comment]
cc: @colhom, @madhusudancs @nikhiljindal Separate issue now.  Since a rebase, I'm seeing incorrect tags being applied to the federation-apiserver and federation-controller-manage images.  This leads to those pods failing to come up (status=ImagePullBackOff).  @nikhiljindal is seeing the same issue.  I haven't debugged very much yet.  Will do now.

Note that the tags on the images uploaded to gcr.io do not match the image tags on the pods created:

```
$ FEDERATION=true E2E_ZONES="us-central1-f" KUBE_GCE_ZONE="us-central1-f" FEDERATION_PUSH_REPO_BASE="gcr.io/google.com/quinton-etcd-testing" KUBERNETES_PROVIDER="gce" federation/cluster/federated-push.sh
Load: /Users/quinton/code/go/src/k8s.io/kubernetes/cluster/../_output/release-stage/server/linux-amd64/kubernetes/server/bin/federation-apiserver.tar
a61215aac3d6: Loading layer 101.1 MB/101.1 MB
Tag: gcr.io/google_containers/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc --> gcr.io/google.com/quinton-etcd-testing/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc
Push: gcr.io/google.com/quinton-etcd-testing/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc
 -> GCR repository detected. Using gcloud
The push refers to a repository [gcr.io/google.com/quinton-etcd-testing/federation-apiserver]
a61215aac3d6: Layer already exists 
5f70bf18a086: Layer already exists 
1834950e52ce: Layer already exists 
e0f90fdaf13303872fcf9ea4c24a62fc: digest: sha256:cf427360cc1500e9b540871038737e771397a8e8105bc59c843ef6d9f013a5bf size: 2663
Remove: gcr.io/google_containers/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc
Untagged: gcr.io/google_containers/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc
Remove: gcr.io/google.com/quinton-etcd-testing/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc
Untagged: gcr.io/google.com/quinton-etcd-testing/federation-apiserver:e0f90fdaf13303872fcf9ea4c24a62fc
Deleted: sha256:2a370c505ba2966cc8b58a39fd739a229597544a5f0b7cb0682118bcb880eee5
Deleted: sha256:d51c10056ab6b6e532fa713bf3abe9d8b08ea4b336c4c19f068f04004694d729
Load: /Users/quinton/code/go/src/k8s.io/kubernetes/cluster/../_output/release-stage/server/linux-amd64/kubernetes/server/bin/federation-controller-manager.tar
7daeea0bb044: Loading layer 49.28 MB/49.28 MB
Tag: gcr.io/google_containers/federation-controller-manager:10a491df9516385bc00493291d4628ca --> gcr.io/google.com/quinton-etcd-testing/federation-controller-manager:10a491df9516385bc00493291d4628ca
Push: gcr.io/google.com/quinton-etcd-testing/federation-controller-manager:10a491df9516385bc00493291d4628ca
 -> GCR repository detected. Using gcloud
The push refers to a repository [gcr.io/google.com/quinton-etcd-testing/federation-controller-manager]
7daeea0bb044: Layer already exists 
5f70bf18a086: Layer already exists 
1834950e52ce: Layer already exists 
10a491df9516385bc00493291d4628ca: digest: sha256:712d40f941e9e25b922c491a10d872700c94411955ccb4f08f430d7687c62801 size: 2681
Remove: gcr.io/google_containers/federation-controller-manager:10a491df9516385bc00493291d4628ca
Untagged: gcr.io/google_containers/federation-controller-manager:10a491df9516385bc00493291d4628ca
Remove: gcr.io/google.com/quinton-etcd-testing/federation-controller-manager:10a491df9516385bc00493291d4628ca
Untagged: gcr.io/google.com/quinton-etcd-testing/federation-controller-manager:10a491df9516385bc00493291d4628ca
Deleted: sha256:fe9ed85ee3f143eb07ef69962858b4502e6b9f55a0ff5a0b217554ec3ca29dc6
Deleted: sha256:8c4412f9968014fd4cbf3265d182ee6c0d660362bc6b900813b63ad62d936def
```

```
$ gcloud docker images gcr.io/google.com/quinton-etcd-testing/federation-controller-manager
REPOSITORY                                                             TAG                                IMAGE ID            CREATED             SIZE
gcr.io/google.com/quinton-etcd-testing/federation-controller-manager   cf6a0fb95915cbbbceca710a499b8344   d5f589df7a70        46 hours ago        50.39 MB
gcr.io/google.com/quinton-etcd-testing/federation-controller-manager   29f7a9da8fdd4d3ea184c92bf2f60719   1af6e3fb3127        4 days ago          50.34 MB
```

```
kubectl --server=https://104.197.162.125 describe pods --namespace=federation-e2e | grep Image: | grep gcr\.io
    Image:		gcr.io/google.com/quinton-etcd-testing/federation-apiserver:v1.3.0-alpha.5.130_a72498e8685b28
    Image:		gcr.io/google.com/quinton-etcd-testing/federation-controller-manager:v1.3.0-alpha.5.130_a72498e8685b28
```

**Comment by quinton-hoole** [comment]
The problem seems to be that the following files, generated via a clean build against upstream/master HEAD last night, contain hexadecimal tags of the form e.g. `e0f90fdaf13303872fcf9ea4c24a62fc`,  whereas the pod manifests are looking for tags of the form 
 `v1.3.0-alpha.5.130_a72498e8685b28`

$ pwd
/Users/quinton/code/go/src/k8s.io/kubernetes/_output/release-stage/server/linux-amd64/kubernetes/server/bin
quinton-macbookpro:bin quinton$ cat *.docker_tag
e0f90fdaf13303872fcf9ea4c24a62fc
10a491df9516385bc00493291d4628ca
d6af5dc588120e8c2f6b428f117db364
578d690956c6673d891046e84137d5e4
d427b03889026070bac283cc36efd3be
fd4f0b7f0da7f6c9a1ec76140fb21662
```

**Comment by quinton-hoole** [comment]
So the tags in the above files are apparently correct:

https://github.com/kubernetes/kubernetes/blob/3b908a538df38d93becda67309321a934138d0f6/build/common.sh#L861

 but they are not being substituted into the pod YAML spec correctly:

https://github.com/kubernetes/kubernetes/blob/3b908a538df38d93becda67309321a934138d0f6/federation/manifests/federation-controller-manager-deployment.yaml#L18

**Comment by quinton-hoole** [comment]
Here's the errant line of code:

hack/e2e-internal/e2e-up.sh:	export FEDERATION_IMAGE_TAG="$(cat "${KUBE_ROOT}/federation/manifests/federated-image.tag")"

**Comment by colhom** [comment]
@quinton-hoole these two values are actually entirely distinct from each other. I think your conflating two steps in the bring-up process.

* **build phase** -- `make-quickrelease` (or equivalent) will leave us the container image as a tarball. `federation-controller-manager.tar.gz`. 

  * The repo will always be `gcr.io/google_containers/federation-controller-manger`
   * The tag will be generated at the conclusion of the docker build and stored in `federation-controller-manager.docker_tag`

* **push phase**  -- `federated-push.sh` will

  * `docker load federation-controller-manager.tar.gz`

    * Oddly enough, `docker load` does not tell you what the `repo:tag` of the image you just loaded is, so you have to figure that on your own. 

       * `repo` is always `gcr.io/google_containers/federation-controller-manager`
       * `tag` is stored in `federation-controller-manager.docker_tag`

  * `docker tag $repo:$tag $FEDERATION_PUSH_REPO_BASE/federation-controller-manger:$VERSION_TAG`
     * `$VERSION_TAG` is simply the semantic version generated for the rest of the ci build artifacts. This makes sense, since the images are also just ci artifacts being pushed to a remote location.

  * `docker push $FEDERATION_PUSH_REPO_BASE/federation-controller-manager:$VERSION_TAG`

**Comment by quinton-hoole** [comment]
OK, I think Colin and I figured this out:

1. ./build/push-ci-build.sh needs to be run manually after e2e.go -build
2. KUBE_GCS_RELEASE_BUCKET needs to be set, otherwise you'll get errors like this: 
    [0607 14:02:25] Copying release artifacts to gs://kubernetes-release/ci/v1.3.0-alpha.5.166+ec4d061f64194e/   AccessDeniedException: 403 Forbidden

**Comment by quinton-hoole** [comment]
I will confirm that this all works properly, and @colhom and/or I will write up and keep current a correct set of instructions.

**Comment by quinton-hoole** [comment]
Confirmed.  My e2e tests now pass.

**Comment by nikhiljindal** [comment]
These are the failures I see on jenkins: http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-federation/

```
[k8s.io] Federation apiserver [Feature:Federation] should allow creation of cluster api objects

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federation-apiserver.go:73 unexpected error in verifying if cluster federation-e2e-gce-us-central1-a is ready: timed out waiting for the condition Expected error: <*errors.errorString | 0xc8200ad060>: { s: "timed out waiting for the condition", } timed out waiting for the condition not to have occurred


[k8s.io] Service [Feature:Federation] should be able to discover a federated service

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federated-service.go:130 Jun 14 10:07:10.209: FEDERATION_NAME environment variable must be set


[k8s.io] Service [Feature:Federation] should be able to discover a non-local federated service

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/federated-service.go:130 Jun 14 10:07:15.987: FEDERATION_NAME environment variable must be set
```

I had verified that apiserver test was passing yesterday evening. Maybe some recent PR broke it?

Also FEDERATION_NAME issue should have been fixed with https://github.com/kubernetes/kubernetes/pull/27333.

Am debugging

**Comment by mml** [comment]
#27390 fixes one problem with the service tests.  The major problem is that the only paths that trigger `syncService()` are via service changes.  None, AFAICT, are via cluster changes, even after #27252.  (Maybe I'm misreading here?)

I can work on a PR for this, but it's definitely more involved, and I don't want to duplicate work if someone's already doing it, so I need an ACK.  From whom I'm not exactly sure.  :-)

**Comment by nikhiljindal** [comment]
@mml Yes please feel free to work on it.
The best way to test your hypothesis is to add a test that fails with the current code.

**Comment by quinton-hoole** [comment]
My understanding was that @mfanjie had fixed that problem in the past few days.  I'll dig out the PR that professes to fix it.

**Comment by quinton-hoole** [comment]
@mml @nikhiljindal The PR I was thinking of is #27028.  I'll re-review now that it actually does what it's supposed to do.

**Comment by quinton-hoole** [comment]
@mml @nikhiljindal I double-checked #27028 and yes, I think you've found a bug.  I'm not sure quite how @mfanjie's manual tests passed. As far as I can see ServiceController.processServiceForCluster() does not in fact invoke ensureDnsRecords(), despite the comment to that effect:

// processServiceForCluster creates or updates service to all registered running clusters,
// update DNS records and update the service info with DNS entries to federation apiserver.
// the function returns any error caught

https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/service/servicecontroller.go#L295

**Comment by nikhiljindal** [comment]
We should add an e2e test case for that.

We should probably compile a list of all test cases that we want. And then each one of use can start picking one.

**Comment by quinton-hoole** [comment]
@nikhiljindal Yes, these are part of the test matrix that @mfanjie sent out, and that @madhusudancs is currently working on.  He is planning to hand over whatever remains to be done to complete the matrix to @mml and @quinton-hoole tomorrow.

**Comment by mfanjie** [comment]
@quinton-hoole my understanding is: 
ensureDNS() should only be invoked when federation service controller catches service updates from underlying k8s clusters, including endpoint update and service update. So it is not invoked in the function processServiceForCluster().
Am I having a misunderstanding here?

**Comment by quinton-hoole** [comment]
@mfanjie I'll have to take a closer look at the code in the morning.  I had (perhaps incorrectly) assumed that processServiceForCluster() was invoked when services in clusters changed. And the comment for that function says that it updates DNS records.  But it does not seem to do that. 

Either way, when underlying services are created, deleted or change status, DNS records may need to be added, deleted or updated.  And when clusters are added or updated, services and DNS records may need to be added or updated.

**Comment by mfanjie** [comment]
@mml cluster changes will trigger service update https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/federation-controller/service/servicecontroller.go#L655
and

**Comment by mfanjie** [comment]
@quinton-hoole "And the comment for that function says that it updates DNS records. But it does not seem to do that."
this comment should be changed, the DNS records update is based on service changes watcher on underlying services.

**Comment by quinton-hoole** [comment]
I'm making fairly good progress doing manual e2e testing, but seeing some curious log messages and behaviour.  I'll note them down here in case anyone else is seeing similar errors:

Of note below is:

1.  At 18:48 4 DNS records were successfully updated, so that appears to work.
2. At 20:07 we detect a new federated service and try to create it in the underlying cluster.  Good.  But this seems to fail due to a misconfigured DNS domain ("example.com").  I'll fix that now.
3. Presumably due to the above failure, the service seems to not be created at all (i.e. not only the DNS record is not created).  Probably OK, but a bit weird. 

```
I0615 18:48:09.849880       1 servicecontroller.go:659] Successfully updated 2 out of 2 DNS records to direct traffic to the updated cluster
I0615 18:48:10.850275       1 servicecontroller.go:645] Detected change in list of cluster names. New  set: map[], Old set: map[federation-e2e-gce-us-central1-f:{}]
I0615 18:48:10.850568       1 servicecontroller.go:659] Successfully updated 4 out of 4 DNS records to direct traffic to the updated cluster
I0615 20:05:10.254312       1 servicecontroller.go:645] Detected change in list of cluster names. New  set: map[federation-e2e-gce-us-central1-f:{}], Old set: map[]
I0615 20:05:10.254468       1 servicecontroller.go:654] New cluster observed federation-e2e-gce-us-central1-f
I0615 20:05:10.257255       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:07:50.264688       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
E0615 20:11:23.972640       1 endpoint_helper.go:88] Failed to sync service: DNS zone example.com not found., put back to service queue
I0615 20:12:50.272983       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:17:50.279530       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
```

**Comment by quinton-hoole** [comment]
@nikhiljindal @madhusudancs @mml @mfanjie FYI ^^

**Comment by quinton-hoole** [comment]
Actually the failure to create the service seems unrelated to the failure to create the DNS record, as it recurs independently every time around the sync loop:

```
I0615 20:12:50.272983       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:17:50.279530       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:22:50.286356       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:27:50.292919       1 servicecontroller.go:417] Service 'e2e-tests-service-d4kwp/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:27:55.299025       1 servicecontroller.go:417] Service 'e2e-tests-service-d4kwp/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:28:05.305809       1 servicecontroller.go:417] Service 'e2e-tests-service-d4kwp/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:28:25.312875       1 servicecontroller.go:417] Service 'e2e-tests-service-d4kwp/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:29:05.319426       1 servicecontroller.go:417] Service 'e2e-tests-service-d4kwp/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:30:25.326059       1 servicecontroller.go:417] Service 'e2e-tests-service-yjdey/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:30:30.332917       1 servicecontroller.go:417] Service 'e2e-tests-service-yjdey/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:30:40.339491       1 servicecontroller.go:417] Service 'e2e-tests-service-yjdey/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:31:00.346206       1 servicecontroller.go:417] Service 'e2e-tests-service-yjdey/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:31:40.352150       1 servicecontroller.go:417] Service 'e2e-tests-service-yjdey/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:33:00.357819       1 servicecontroller.go:417] Service 'e2e-tests-service-9p7ro/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:35:40.364418       1 servicecontroller.go:417] Service 'e2e-tests-service-9p7ro/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
I0615 20:40:40.373982       1 servicecontroller.go:417] Service 'e2e-tests-service-9p7ro/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new
```

**Comment by mml** [comment]
I'm having success with some DNS names but not others.  For example

```
Server:    10.0.0.10
Address 1: 10.0.0.10 kube-dns.kube-system.svc.cluster.local

Name:      federated-service
Address 1: 10.0.254.162 federated-service.e2e-tests-service-ev6w0.svc.cluster.local
```

and

```
Server:    10.0.0.10
Address 1: 10.0.0.10 kube-dns.kube-system.svc.cluster.local

Name:      federated-service.e2e-tests-service-ev6w0
Address 1: 10.0.254.162 federated-service.e2e-tests-service-ev6w0.svc.cluster.local
```

and

```
Server:    10.0.0.10
Address 1: 10.0.0.10 kube-dns.kube-system.svc.cluster.local

Name:      federated-service.e2e-tests-service-ev6w0.svc.cluster.local.
Address 1: 10.0.254.162 federated-service.e2e-tests-service-ev6w0.svc.cluster.local
```

but `federated-service.e2e-tests-service-ev6w0.federation` fails.

Still digging in to that, but I note that having all these be a single test makes it really hard to debug what failed, or even to note that three of the DNS entries passed.

**Comment by nikhiljindal** [comment]
Sent https://github.com/kubernetes/kubernetes/pull/27504 to dump federation apiserver and controller manager logs on failure which should help in debugging federation failures on jenkins.

I am right now looking at federation-apiserver test. Its failing on jenkins. Works fine locally.

**Comment by mfanjie** [comment]
@quinton-hoole FYI, this log indicates service controller are trying to create new service on k8s cluster federation-e2e-gce-us-central1-f, and repeated logs might be led by creation failure.
```
I0615 20:12:50.272983       1 servicecontroller.go:417] Service 'e2e-tests-service-3s4ko/federated-service' is not found in cluster federation-e2e-gce-us-central1-f, trying to create new

```

**Comment by quinton-hoole** [comment]
This loos like a bug.   Namespace being dropped?  I'll send in a fix shortly.

```
[BeforeEach] DNS
  /Users/quinton/code/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/federated-service.go:136
STEP: Creating federated service "federated-service" in namespace "e2e-tests-service-40rio"
Jun 17 11:43:48.605: INFO: Trying to create service "federated-service" in namespace ""
```

**Comment by nikhiljindal** [comment]
federation apiserver e2e test was a lot more green over the weekend. It seems to have regressed again

![fed-apiserver-flake](https://cloud.githubusercontent.com/assets/10199099/16246705/3727fb8e-37bb-11e6-8944-e1c9942b8452.png)

**Comment by colhom** [comment]
We should probably wrap the underlying `kube-up` invocations in a retry loops? That would probably take care of most of the failures.

**Comment by quinton-hoole** [comment]
I think these issues are all resolved now, and this issue has served it's purpose. Closing.
